skbio/metadata/_mixin.py

```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
         """
         return (self._positional_metadata is not None and
                 len(self.positional_metadata.columns) > 0)
+
+
+class IntervalMetadataMixin(metaclass=abc.ABCMeta):
+    def __init__(self, interval_metadata=None):
+        raise NotImplementedError
+
+    def _init_(self, interval_metadata=None):
+        self.interval_metadata = interval_metadata
+
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+
+        >>> from skbio import DNA
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> intvl = im.add([(0, 3)], metadata={'gene': 'sagA'})
+        >>> seq = DNA('ACGT', interval_metadata=im)
+        >>> seq
+        DNA
+        --------------------------
+        Interval metadata:
+            1 interval features
+        Stats:
+            length: 4
+            has gaps: False
+            has degenerates: False
+            has definites: True
+            GC-content: 50.00%
+        --------------------------
+        0 ACGT
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata()
```

Comment:
We'll need the `IntervalMetadata` constructor to accept a length, and validate that interval features all fall within that coordinate space when they are added. This is necessary to ensure that interval metadata spans the length of the sequence, similar to how the number of rows in positional metadata (`pd.DataFrame`) match the sequence length.

This doesn't have to be called "length", since that could be confused with the number of interval features in the interval metadata. I'm not sure what to call it, do you have any ideas?

Once `IntervalMetadata` has a "length", `IntervalMetadata._reverse` can be renamed `.reverse()` and will no longer need to accept `length` as a parameter (since it will be instance state).

This mixin will need an abstract method similar to `_positional_metadata_axis_len_` that objects must implement to define their length. Probably makes sense to call it `_interval_metadata_axis_len_` (or some variation if we have a better name than "length").